rudder-core/src/main/scala/com/normation/rudder/services/policies/JavascriptEngine.scala

```diff
@@ -0,0 +1,708 @@
+
+      }
+    case AixHash =>
+      new JsLibPassword() {
+        /// method accessible from JS
+        def md5    (s: String): String = super.aixMd5(s)
+        def sha256 (s: String): String = super.aixSha256(s)
+        def sha512 (s: String): String = super.aixSha512(s)
+
+        def md5    (s: String, salt: String): String = super.aixMd5(s, salt)
+        def sha256 (s: String, salt: String): String = super.aixSha256(s, salt)
+        def sha512 (s: String, salt: String): String = super.aixSha512(s, salt)
+
+      }
+  }
+
+  // with the getter, it will be accessed with rudder.password...
+  def getPassword = password
+}
+
+sealed trait JsRudderLibBinding { def bindings: Bindings }
+
+object JsRudderLibBinding {
+
+  import java.util.{ HashMap => JHMap }
+  import javax.script.SimpleBindings
+
+  private[this] def toBindings(k: String, v: JsRudderLibImpl): Bindings = {
+    val m = new JHMap[String, Object]()
+    m.put(k, v)
+    new SimpleBindings(m)
+  }
+
+  /*
+   * Be carefull, as bindings are mutable, we can't have
+   * a val for bindings, else the same context is shared...
+   */
+  final object Aix extends JsRudderLibBinding {
+    def bindings = toBindings("rudder", new JsRudderLibImpl(AixHash))
+  }
+
+  final object Crypt extends JsRudderLibBinding {
+    def bindings = toBindings("rudder", new JsRudderLibImpl(CryptHash))
+  }
+}
+
+
+/**
+ * This class provides the Rhino (java 7) or Longhorn (Java 8 & up)
+ * java script engine.
+ *
+ * It allows to eval parameters in directives which are starting
+ * with $eval.
+ *
+ */
+final object JsEngineProvider {
+
+
+  /**
+   * Initialize a new JsEngine with the correct bindings.
+   *
+   * Not all Libs (bindings in JSR-223 name) can't be provided here,
+   * because we want to make them specific to each eval
+   * (i.e: different eval may have to different bindings).
+   * So we just provide eval-indep lib here, but we don't
+   * have any for now.
+   */
+  def withNewEngine[T](feature: FeatureSwitch)(script: JsEngine => Box[T]): Box[T] = {
+    feature match {
+      case FeatureSwitch.Enabled  =>
+        SandboxedJsEngine.sandboxed { engine => script(engine) }
+      case FeatureSwitch.Disabled =>
+        script(DisabledEngine)
+    }
+  }
+
+}
+
+sealed trait JsEngine {
+  /**
+   *
+   * Parse a value looking for EVAL keyword.
+   * Return the correct NodeContextualizedValue type.
+   *
+   * Note that the eval result is ALWAY cast to
+   * string. So computation, object, etc must be
+   * correctly defined by the user to get and
+   * interesting value.
+   */
+  def eval(variable: Variable, lib: JsRudderLibBinding): Box[Variable]
+}
+
+/*
+ * Our JsEngine.
+ */
+final object JsEngine {
+  // Several evals: one default and one JS (in the future, we may have several language)
+  final val DEFAULT_EVAL = "eval"
+  final val EVALJS = "evaljs"
+
+  final val default_pattern = """(?ms)(.*)\$\{eval\s+(.*)}(.*)""".r
```

Comment:
Parsing ${evaljs ...} with regex won't work in intersting fashion, especially with rudder variables and js blocks. 
In all case, it will have a massive impact on generation time - we switch from a "startWith" method, dicidable in constant(EVALJS.size) time, to an unbound one - at least o(n), but with extrem backtracking bad case, see stackoferflow recent outage. 

Finally, I don't even think we should have {}. It add visual clutters for no reason, because really, we want that the whole input is a script or not. Supporting multiple script need to have a ROBUST input parser, understanding each language to know if the closing } is its or ours - something I don't even want to start coding. And if it is to have part of the input "as is", here you go : javascript is able to manage strings !

So to sum-up: 
- I don't thing we should have {} letting the user think it can have several eval in an input ; 
- I don't thing we should autorise $evaljs at an other place than the strict begining of the field.